alphadia/workflow/peptidecentric/extraction_handler.py

```diff
@@ -390,17 +440,144 @@ def _select_candidates(
+            fragment_mz_column=self._column_name_handler.get_fragment_mz_column(),
+        )
+
+        candidates = candidates_to_ng(candidates_df, self.cycle_len)
+
+        scoring_params = ScoringParameters()
+        scoring_params.update(
+            {
+                "top_k_fragments": 99,
+                "mass_tolerance": 7.0,
+            }
+        )
+
+        candidate_features = PeakGroupScoring(scoring_params).score(
+            dia_data_ng, speclib_ng, candidates
+        )
+
+        features_df = to_features_df(candidate_features, spectral_library)
+
+        # features_df["elution_group_idx"] = 1 # search_parameter_optimization -> _optlock.update_with_extraction
+        # features_df["decoy"] = 0 # search_parameter_optimization -> fdr_manager.fit_predict
+        # features_df["mz_observed"] = 1 # perform_fdr -> fragment_competition
+        # features_df["rt_observed"] = 1 # perform_fdr -> fragment_competition
+        # features_df["channel"] = 0 # perform_fdr -> keep_best
+        # features_df["proteins"] = 1000 # log_precursor_df
+
+        return features_df, None
+
+    def quantify_ng(
+        self,
+        candidates_df: pd.DataFrame,
+        features_df: pd.DataFrame,
+        dia_data: "DiaDataNG",  # noqa: F821
+        spectral_library: SpecLibFlat,
+        fdr_manager: FDRManager,
+        classifier_version: int,
+    ) -> tuple[pd.DataFrame, pd.DataFrame]:
+        # TODO needs to be stored
+        speclib_ng = speclib_to_ng(
+            spectral_library,
+            rt_column=self._column_name_handler.get_rt_column(),
+            mobility_column=self._column_name_handler.get_mobility_column(),
+            precursor_mz_column=self._column_name_handler.get_precursor_mz_column(),
+            fragment_mz_column=self._column_name_handler.get_fragment_mz_column(),
+        )
+
+        # TODO: why not use candidate_hash here?
+        features_df["precursor_idx_rank"] = (
+            features_df["precursor_idx"].astype(str)
+            + "_"
+            + features_df["rank"].astype(str)
+        )
+        candidates_df["precursor_idx_rank"] = (
+            candidates_df["precursor_idx"].astype(str)
+            + "_"
+            + candidates_df["rank"].astype(str)
+        )
+        # TODO: think about how to making filtering nice XXX
+
+        # apply FDR to PSMs
+        precursor_fdr_df = fdr_manager.fit_predict(
+            features_df,
+            decoy_strategy="precursor",  # TODO support channel_wise
+            competetive=False,  # self._config["fdr"]["competetive_scoring"],
+            df_fragments=None,  # TODO: support fragments_df,
+            version=classifier_version,
+        )
+        precursor_fdr_df = precursor_fdr_df[
+            precursor_fdr_df["qval"] <= self._config["fdr"]["fdr"]
+        ]
+
+        # filter2: candidates by precursors
+        candidates_filtered = candidates_df[
```

Comment:
let's collect a number of stuff we want to move down to the Rust core.

<img width="640" height="761" alt="Image" src="https://github.com/user-attachments/assets/c7f5a360-d8a1-48c0-985b-52902a839a92" />